cdap-watchdog/src/main/java/io/cdap/cdap/logging/gateway/handlers/LogHttpHandler.java

```diff
@@ -115,6 +128,40 @@ public void getRunIdLogs(HttpRequest request, HttpResponder responder,
         escape, filterStr, runRecord, format, suppress);
   }
 
+  @POST
+  @Path("/namespaces/{namespace-id}/apps/{app-id}/{program-type}/{program-id}/runs/{run-id}/logs/classify")
+  public void classifyRunIdLogs(HttpRequest request, HttpResponder responder,
+      @PathParam("namespace-id") String namespaceId,
+      @PathParam("app-id") String appId, @PathParam("program-type") String programType,
+      @PathParam("program-id") String programId,
+      @PathParam("run-id") String runId) throws Exception {
+    ensureVisibilityOnProgram(namespaceId, appId, programType, programId);
+    ProgramType type = ProgramType.valueOfCategoryName(programType);
+    ProgramReference programRef = new ProgramReference(namespaceId, appId, type, programId);
+    RunRecordDetail runRecord = getRunRecordMeta(programRef, runId);
+    if (runRecord.getStatus() != ProgramRunStatus.FAILED) {
+      throw new IllegalArgumentException("Classification is only supported for failed runs");
+    }
+    LoggingContext loggingContext = LoggingContextHelper.getLoggingContextWithRunId(programRef,
+        runId, runRecord.getSystemArgs());
+
+    Filter filter = FilterParser.parse("loglevel=ERROR");
+    ReadRange readRange = new ReadRange(0, System.currentTimeMillis(),
+        LogOffset.INVALID_KAFKA_OFFSET);
+    readRange = adjustReadRange(readRange, runRecord, true);
+    try {
+      // the iterator is closed by the BodyProducer passed to the HttpResponder
+      CloseableIterator<LogEvent> logIter = logReader.getLog(loggingContext,
```

Comment:
Use try with resources. The iterator must be closed after we exit `classifier.classify`

Reply:
Done.